fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvPartitionCompensator.java

```diff
@@ -463,6 +479,58 @@ public static OptExpression getMvTransparentPlan(MaterializationContext mvContex
         return result;
     }
 
+    /**
+     * In some cases, need add cast project to make sure the output columns are the same as expectOutputColumns.
+     * <p>
+     * eg: table t1: k1 date, v1 int, v2 char(20)
+     * mv: create mv mv1 as select k1, v1, v2 from t1.
+     * mv's schema will be: k1 date, v1 int, v2 varchar(20)
+     * </p>
+     * It needs to add a cast project in generating the union operator, which is the same as
+     * {@link com.starrocks.sql.optimizer.transformer.RelationTransformer#processSetOperation(SetOperationRelation)}
+     * </p>
+     * @param columnRefFactory column ref factory to generate the new query column ref
+     * @param optExpression the original opt expression
+     * @param outputColumns the original output columns of the opt expression
+     * @param expectOutputColumns the expected output columns
+     * @return the new opt expression and the new output columns if it needs to cast, otherwise return the original
+     */
+    private static Pair<OptExpression, List<ColumnRefOperator>> addCastProjectIfNeeded(
+            ColumnRefFactory columnRefFactory,
+            OptExpression optExpression,
+            List<ColumnRefOperator> outputColumns,
+            List<ColumnRefOperator> expectOutputColumns) {
+        if (outputColumns.size() != expectOutputColumns.size()) {
+            return null;
+        }
+        int len = outputColumns.size();
+        Map<ColumnRefOperator, ScalarOperator> projections = Maps.newHashMap();
+        List<ColumnRefOperator> newChildOutputs = new ArrayList<>();
+        boolean isNeedCast = false;
+        for (int i = 0; i < len; i++) {
```

Comment:
can we use  a more succinct method?
for an example:

insert cast for each output column, then invoke ReduceCastRule to  each output column;

Reply:
For now, should not add one more rule yet, because only mv union rewrite can cause union children's type are not matched. for query, it will be matched in transformer.